### ISSUE TITLE
JWT decoding failure when upgrading from 7.8.1 to 7.9.0

### ISSUE THREAD

**Issue by jsm174** [jhipster/generator-jhipster#19309]
##### **Overview of the issue**

I was working on upgrading a monolith oauth2 app from 7.8.1 to 7.9.1.

We currently use Keycloak 15.1.1.

When trying to call any services from our React app I am getting 401s.

After examining the response headers, I see:

```WWW-Authenticate: Bearer error="invalid_token", error_description="An error occurred while attempting to decode the Jwt: timestamps must be of type Instant: java.util.Date", error_uri="https://tools.ietf.org/html/rfc6750#section-3.1"```

I manually decoded the JWT and `iat` and `exp` seem fine:

```
  "exp": 1659374140,
  "iat": 1659373840,
```

It looks like something in `nimbus` is converting them to `java.util.Date` instead of `Instant`.

Is anyone else seeing this issue?

##### **Motivation for or Use Case**

This should work out of the box.

##### **Suggest a Fix**

As a work around, I manually converted the `iat` and `exp` claims to `Instant` in `CustomClaimConverter`:

```
    public Map<String, Object> convert(Map<String, Object> claims) {
    	LinkedHashMap<String, Object> tempClaims = new LinkedHashMap<String, Object>();
    	for (String key : claims.keySet()) {
    	    Object value = claims.get(key);
    	    if (key.equals("exp") || key.equals("iat")) {
    	        value = ((Date) value).toInstant();
    	    }
    	    tempClaims.put(key, value);
    	}
        // Only look up user information if identity claims are missing
        if (tempClaims.containsKey("given_name") && tempClaims.containsKey("family_name")) {
            return tempClaims;
        }
        Map<String, Object> convertedClaims = this.delegate.convert(tempClaims);
```

##### **JHipster Version(s)**

```
test-app@0.0.0 /Users/jmillard/git/test-app
└── generator-jhipster@7.9.0

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "applicationIndex": 0,
    "applicationType": "monolith",
    "authenticationType": "oauth2",
    "baseName": "TestApp",
    "blueprints": [],
    "buildTool": "gradle",
    "cacheProvider": "ehcache",
    "clientFramework": "angularX",
    "clientPackageManager": "npm",
    "clientTheme": "none",
    "clientThemeVariant": "",
    "creationTimestamp": 1609459200000,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "dtoSuffix": "DTO",
    "enableGradleEnterprise": false,
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "entities": ["LastSeq"],
    "entitySuffix": "",
    "gradleEnterpriseHost": "",
    "jhiPrefix": "jhi",
    "jhipsterVersion": "7.9.0",
    "languages": ["en"],
    "lastLiquibaseTimestamp": 1609459260000,
    "messageBroker": false,
    "nativeLanguage": "en",
    "otherModules": [],
    "packageFolder": "com/testapp",
    "packageName": "com.testapp",
    "pages": [],
    "prodDatabaseType": "postgresql",
    "reactive": false,
    "searchEngine": false,
    "serverPort": "8080",
    "serviceDiscoveryType": false,
    "skipCheckLengthOfIdentifier": false,
    "skipClient": true,
    "skipFakeData": true,
    "skipUserManagement": true,
    "testFrameworks": [],
    "websocket": false,
    "withAdminUi": true
  }
}

</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity LastSeq {
  id UUID
  startSeq Long required
  endSeq Long required
  createdAt Instant
  lastUpdate Instant
}
paginate LastSeq with pagination

</pre>
</details>

##### **Environment and Tools**

openjdk version "18.0.1.1" 2022-04-22
OpenJDK Runtime Environment Homebrew (build 18.0.1.1+0)
OpenJDK 64-Bit Server VM Homebrew (build 18.0.1.1+0, mixed mode, sharing)

git version 2.36.1

node: v16.16.0

npm: 8.11.0

Docker version 20.10.17, build 100c701

Docker Compose version v2.6.1

##### **Browsers and Operating System**

MacOS 12.4
Chome 103.0.5060.134

- [X] Checking this box is mandatory (this is just to show you read everything)

**Comment by budthapa** [comment]
I am also getting the exact same error (401) and message after upgrading from 7.8.1 to 7.9.0.
We are using the Keycloak version: 15.0.2. Frontend is vue app

**Comment by atomfrede** [comment]
We test (and support) keycloak 18 with 7.9.0 out of box. Can you try it with that version? Maybe we need to put the workaround into upgrade notices.

CC @mraible Are you aware of something which might cause this behaviour? I didn't had the issue when trying a fresh app from main with provided keycloak version.

**Comment by gmarziou** [comment]
Maybe you should compare dependencies versions, maybe the jwt lib has introduced a regression

gradlew -q dependencies

**Comment by jsm174** [comment]
@atomfrede - thanks. Unfortunately, I can't upgrade Keycloak in our production environment at this time.

I was able to setup my local environment with Keycloak 18 and am still seeing this issue. 

IMHO, I don't think this is a Keycloak issue --  I've checked JWTs from other sites, and `iat` and `exp` are all the same format.

For additional context, our website that connects to our jhipster services is a standalone react app.  It uses `oidc-client-ts` and `react-oidc-context` to handle tokens. (ie, similar to `keycloak-js`).  

An out of the box jhipster app doesn't use the authorization request header.

**Comment by atomfrede** [comment]
Sure updating keycloak is no option. So agree looks like a jhipster bug. Maybe difference between maven and gradle? Will try to reproduce it.

**Comment by atomfrede** [comment]
There is the `MappedJwtClaimSetConverter`, configured like below. So it looks like  either the converter is not using the defined conversion or it doesn't work (which at least the code does not allow as it would throw an exception).

```
   public static MappedJwtClaimSetConverter withDefaults(Map<String, Converter<Object, ?>> claimTypeConverters) {
        Assert.notNull(claimTypeConverters, "claimTypeConverters cannot be null");
        Converter<Object, ?> stringConverter = getConverter(STRING_TYPE_DESCRIPTOR);
        Converter<Object, ?> collectionStringConverter = getConverter(TypeDescriptor.collection(Collection.class, STRING_TYPE_DESCRIPTOR));
        Map<String, Converter<Object, ?>> claimNameToConverter = new HashMap();
        claimNameToConverter.put("aud", collectionStringConverter);
        claimNameToConverter.put("exp", MappedJwtClaimSetConverter::convertInstant);
        claimNameToConverter.put("iat", MappedJwtClaimSetConverter::convertInstant);
        claimNameToConverter.put("iss", MappedJwtClaimSetConverter::convertIssuer);
        claimNameToConverter.put("jti", stringConverter);
        claimNameToConverter.put("nbf", MappedJwtClaimSetConverter::convertInstant);
        claimNameToConverter.put("sub", stringConverter);
        claimNameToConverter.putAll(claimTypeConverters);
        return new MappedJwtClaimSetConverter(claimNameToConverter);
    }
```

**Comment by atomfrede** [comment]
@jsm174 Can you check which nimbus version is resolved? I can't reproduce it right now with either maven nor gradle (but I am also not getting into the convert method with the debugger).

**Comment by jsm174** [comment]
Not sure if I'm doing this correctly, but I wiped `~/.gradle` and `.gradle`

```
./gradlew -q dependencies | grep -i nimbus
|    |    \--- com.nimbusds:oauth2-oidc-sdk:9.35
|    |         +--- com.nimbusds:content-type:2.2
|    |         +--- com.nimbusds:lang-tag:1.6
|    |         \--- com.nimbusds:nimbus-jose-jwt:9.22
|         \--- com.nimbusds:nimbus-jose-jwt:9.22 (*)
|    |    \--- com.nimbusds:oauth2-oidc-sdk:9.35
|    |         +--- com.nimbusds:content-type:2.2
|    |         +--- com.nimbusds:lang-tag:1.6
|    |         \--- com.nimbusds:nimbus-jose-jwt:9.22
|         \--- com.nimbusds:nimbus-jose-jwt:9.22 (*)
|    |    \--- com.nimbusds:oauth2-oidc-sdk:9.35
|    |         +--- com.nimbusds:content-type:2.2
|    |         +--- com.nimbusds:lang-tag:1.6
|    |         \--- com.nimbusds:nimbus-jose-jwt:9.22
|         \--- com.nimbusds:nimbus-jose-jwt:9.22 (*)
|    |    \--- com.nimbusds:oauth2-oidc-sdk:9.35
|    |         +--- com.nimbusds:content-type:2.2
|    |         +--- com.nimbusds:lang-tag:1.6
|    |         \--- com.nimbusds:nimbus-jose-jwt:9.22
|         \--- com.nimbusds:nimbus-jose-jwt:9.22 (*)
|    |    \--- com.nimbusds:oauth2-oidc-sdk:9.35
|    |         +--- com.nimbusds:content-type:2.2
|    |         +--- com.nimbusds:lang-tag:1.6
|    |         \--- com.nimbusds:nimbus-jose-jwt:9.22
|         \--- com.nimbusds:nimbus-jose-jwt:9.22 (*)
|    |    \--- com.nimbusds:oauth2-oidc-sdk:9.35
|    |         +--- com.nimbusds:content-type:2.2
|    |         +--- com.nimbusds:lang-tag:1.6
|    |         \--- com.nimbusds:nimbus-jose-jwt:9.22
|         \--- com.nimbusds:nimbus-jose-jwt:9.22 (*)
```

**Comment by atomfrede** [comment]
Thanks thats fine. So 9.22 what it have too. And you are "just" e.g. opening an entity page in the ui to get that error? Using a plain jhipster monolith?

**Comment by jsm174** [comment]
(Just updated above with 7.8.1 deps.)

No, we have a standalone react app that uses something similar to `keycloak.js` to handle access / refresh tokens. We then use that in the authorization header when calling the services in the jhipster monolith.

**Comment by jsm174** [comment]
<img width="1799" alt="Screen Shot 2022-08-05 at 8 22 43 AM" src="https://user-images.githubusercontent.com/1197137/183076475-05cf9ff9-372e-40ae-8a8f-b325c70e0707.png">

Here's a little bit of the stack trace right at the `CustomClaimConverter`.

**Comment by atomfrede** [comment]
I don't stop in this methods when using the generated jhipster ui. Need to check, but yes it looks a little strange, definitly not an instant in the token

**Comment by JeroenAdam** [comment]
Same here for a fresh app generated by Jhipster v7.9.2
I used the same JDL from a running app which was generated by Jhipster v7.4.1, no such issue with that app.
Keycloak server v16.1.1

How to reproduce in 2 easy steps with Postman:

Obtain an access token via POST request to /realms/jhipster/protocol/openid-connect/token
These 5 keys in body: client_id client_secret username password
grant_type (value must be password)

Send a get request to any endpoint with Authorization header containing the Bearer token
The header response will show this:
www-authenticate: "Bearer error="invalid_token", error_description="An error occurred while attempting to decode the Jwt: timestamps must be of type Instant: java.util.Date", error_uri="https://tools.ietf.org/html/rfc6750#section-3.1""

**Comment by atomfrede** [comment]
Thanks for the reproducer. Will have a look what's changed between 7.8.x and 7.9.x

**Comment by rgordeev** [comment]
Same problem. Are there any ideas, what the reason?

**Comment by atomfrede** [comment]
Not sure, it seems like a change in a 3rd party dependency (com.nimbusds:nimbus-jose-jwt). Will need to take a deeper look.

**Comment by jsm174** [comment]
> @jsm174 Are you doing authentication in your React app and passing an access token to your APIs? If so, this is not the default setup that we use in JHipster.

Correct. It is not the default setup.

**Comment by mraible** [comment]
I wonder if it's because of the access token you're obtaining through Postman? The backend of a JHipster app acts as a resource server by default. That's how we support our Ionic and React Native blueprints. The authentication happens on the mobile app with Keycloak, then the access token is used to communicate with the API. The API is smart enough to look up the user's details (first name, last name, etc) if they're not passed as claims in the access token.

What library are you using to do authentication in your React app?

Part of the reason we architected things the way they are is that 1) it's more secure because you're not storing access tokens on the client and 2) it's easier to maintain because we don't need to integrate an OIDC library into each frontend framework.

**Comment by JeroenAdam** [comment]
FYI, beside Postman, I also have the same error when trying to populate some data via [Node-Red ](https://nodered.org/). The error already occurs on the first GET request to obtain the XSRF-TOKEN, normally that works (with a valid Bearer token), and a second request follows with JSON data (eg. adding some entity).

**Comment by mraible** [comment]
Regardless, I don't think this is a bug since everything works just fine when you use Ionic or React Native and call JHipster's APIs. 

If you'd like to create a PR to fix things so it works for you, I can test with Ionic/React Native and see if everything still works.

**Comment by jsm174** [comment]
This approach has worked in JHipster 6+. I have been religious at updating our websites every JHipster version.

If you look at the workaround I posted in the issue that is how I am getting around it.

> The authentication happens on the mobile app with Keycloak, then the access token is used to communicate with the API. The API is smart enough to look up the user's details (first name, last name, etc) if they're not passed as claims in the access token.

This is exactly what we are doing.

**Comment by mraible** [comment]
To try and figure out if something is broken in 7.9, I cloned the example from my [Ionic tutorial](https://developer.okta.com/blog/2022/05/12/ionic-angular-jhipster). It uses 7.8.1, so I figured it'd be best to make sure that still works first.

```
git clone https://github.com/oktadev/okta-jhipster-ionic-example.git
cd okta-jhipster-ionic-example/backend
npm run ci:e2e:prepare # uses keycloak:16.1.0
./mvnw

# open a new terminal
cd okta-jhipster-ionic-example/ionic-app
npm install
npm start
```

It does, so I tried upgrading the backend using `npx jhipster upgrade`. After fixing conflicts, I stopped all the Docker containers and restarted them. 

```
npm run ci:e2e:prepare # uses keycloak:18.0.0
./mvnw
```

This spits out some errors:

```
[INFO] --- frontend-maven-plugin:1.12.1:npm (npm install) @ flickr-2 ---
[INFO] Running 'npm install' in /Users/mraible/Downloads/okta-jhipster-ionic-example/backend
[INFO] npm ERR! code ERESOLVE
[INFO] npm ERR! ERESOLVE could not resolve
[INFO] npm ERR! 
[INFO] npm ERR! While resolving: flickr-2@0.0.1-SNAPSHOT
[INFO] npm ERR! Found: @testing-library/react@12.1.4
[INFO] npm ERR! node_modules/@testing-library/react
[INFO] npm ERR!   dev @testing-library/react@"13.3.0" from the root project
[INFO] npm ERR! 
[INFO] npm ERR! Could not resolve dependency:
[INFO] npm ERR! dev @testing-library/react@"13.3.0" from the root project
[INFO] npm ERR! 
[INFO] npm ERR! Conflicting peer dependency: react@18.2.0
[INFO] npm ERR! node_modules/react
[INFO] npm ERR!   peer react@"^18.0.0" from @testing-library/react@13.3.0
[INFO] npm ERR!   node_modules/@testing-library/react
[INFO] npm ERR!     dev @testing-library/react@"13.3.0" from the root project
[INFO] npm ERR! 
[INFO] npm ERR! Fix the upstream dependency conflict, or retry
[INFO] npm ERR! this command with --force, or --legacy-peer-deps
[INFO] npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

I tried `rm -rf node_modules package-lock.json` followed by `npm install`. 

```
npm ERR! Could not resolve dependency:
npm ERR! peer react@"^16.2.0 || ^17.0.0" from react-images@1.2.0-beta.7
npm ERR! node_modules/react-images
npm ERR!   react-images@"^1.2.0-beta.7" from the root project
```

Since there's not a new version of `react-images` that works with React 18, I tried to solve things with `npm install --force`. That didn't work, so I removed `react-images` as a dependency.

I'm able to log in to my backend just fine, but I'm not able to log in to the Ionic app. There are no errors either. 

I tried your workaround:

```java
public Map<String, Object> convert(Map<String, Object> claims) {
    // workaround for Keycloak 18
    LinkedHashMap<String, Object> tempClaims = new LinkedHashMap<>();
    for (String key : claims.keySet()) {
        Object value = claims.get(key);
        if (key.equals("exp") || key.equals("iat")) {
            value = ((Date) value).toInstant();
        }
        tempClaims.put(key, value);
    }
    // Only look up user information if identity claims are missing
    if (claims.containsKey("given_name") && claims.containsKey("family_name")) {
        return tempClaims;
    }
    Map<String, Object> convertedClaims = this.delegate.convert(tempClaims);
    ...
}
```

This does solve the problem, so I'll create a PR. Thanks!

**Comment by jsm174** [comment]
Thanks for the PR. For clarity, this is not a Keycloak 18 issue.  I don't even think its a Keycloak issue.

**Comment by mraible** [comment]
FWIW, I converted my backend to be reactive and everything works with Ionic. No workaround is needed.

**Comment by mraible** [comment]
> For clarity, this is not a Keycloak 18 issue. I don't even think it's a Keycloak issue.

I tried creating a JHipster + Ionic app with [blog-oauth2.jdl](https://github.com/jhipster/jdl-samples/blob/main/blog-oauth2.jdl) and Okta. This workaround is not needed.

**Comment by jsm174** [comment]
Can you get a copy of your access token and decode it, and see what the `iat` and `exp` values are?

ie, are they values similar to:

```
  "exp": 1659374140,
  "iat": 1659373840,
```

**Comment by mraible** [comment]
Here's a sample access token:

```
eyJraWQiOiJXWGFJUTNIZ1dQY2hFMndtRDNYclRJYndpS3hNakdWX2hqRjR3T0tDYmk4IiwiYWxnIjoiUlMyNTYifQ.eyJ2ZXIiOjEsImp0aSI6IkFULi1ISzl0dEZ5NV82ZHk4NjFOeFpiODVlSW1HZEE2VFVQRXdkaEh4WEN2VW8iLCJpc3MiOiJodHRwczovL2Rldi0xNzcwMDg1Ny5va3RhLmNvbS9vYXV0aDIvZGVmYXVsdCIsImF1ZCI6ImFwaTovL2RlZmF1bHQiLCJpYXQiOjE2NjA2ODk4NTYsImV4cCI6MTY2MDY5MzQ1NiwiY2lkIjoiMG9hNjZzZWwzNGZBa053alI1ZDciLCJ1aWQiOiIwMHUzdzQ3eWhtMVhaWUNKWDVkNyIsInNjcCI6WyJwcm9maWxlIiwib3BlbmlkIl0sImF1dGhfdGltZSI6MTY2MDY4OTg1NCwic3ViIjoibWF0dEByYWlibGVkZXNpZ25zLmNvbSIsImdyb3VwcyI6WyJFdmVyeW9uZSIsIlJPTEVfVVNFUiIsIlJPTEVfQURNSU4iXX0.Kw6GVbpzLcjLHe8KKTVg39kIAw-HQCHTy_bduAeMegOBwsUHhrwD8pWTNbC-C5fYqgIh_Oe9E-zzcJL88fIp7RGdlYN0cNWcgQk3MUju-qBoRpFv5rSNo3VoqM4uHXPLrmPAQhL4QPCTJASsSRUyGoXOkH4G8hWnFKEbzpKhSfrwXrB0bjwD_GrFgm6Ff1yVUlZVj-0y0ntURMg7aN-HkCCNxKsktxCS8It4tep56TJlBWQrueuiHpTZfd-zJuSYpKoiygY4PJxH7L7m3an-M50v3sXfn08FXSQUWYN35dUMExqdOZ_7yr_MWYwm348_mUrS1evsS0FPnsrLGAhk_A
```

If I decode it at jwt.io, it says:

```
  "iat": 1660689856,
  "exp": 1660693456,
```

**Comment by jsm174** [comment]
Thanks. I took your token and attempted to get into the `CustomClaimConverter`:

```
value: [api://default], class=java.util.ArrayList
value: ******, class=java.lang.String
value: ["profile","openid"], class=com.nimbusds.jose.shaded.json.JSONArray
value: ****@***.com, class=java.lang.String
value: 1, class=java.lang.Long
value: 1660689854, class=java.lang.Long
value: **************, class=java.lang.String
value: ["Everyone","ROLE_USER","ROLE_ADMIN"], class=com.nimbusds.jose.shaded.json.JSONArray
value: Tue Aug 16 19:44:16 EDT 2022, class=java.util.Date
value: Tue Aug 16 18:44:16 EDT 2022, class=java.util.Date
value: ******, class=java.lang.String
value: ******, class=java.lang.String
```

`iat` and `exp` are `java.util.Date` in yours too. 

The difference between your token from Okta and my token from Keycloak, is that mine includes `given_name` and `family_name`.  

That alters the flow in `CustomClaimConverter`

```
        if (claims.containsKey("given_name") && claims.containsKey("family_name")) {
            return claims;
        }
```

Could you try to add `given_name` and `family_name` to your okta claims, and try again?

**Comment by mraible** [comment]
If I try adding given and family name claims, you are correct, authentication does fail. If you have these values in your access token, you really don't need the `CustomClaimConverter` in your project. However, putting identity information in access tokens is also a bad practice.

**Comment by mshima** [comment]
@mraible can you test https://github.com/jhipster/generator-jhipster/pull/19454?

**Comment by mraible** [comment]
@mshima Good eye! Such a small change and it totally solves the problem. Well done! 👊

**Comment by gmarziou** [comment]
> However, putting identity information in access tokens is also a bad practice.

Why @mraible ?

**Comment by mraible** [comment]
@gmarziou Because access tokens are only meant for authorization and can be opaque (a random string of characters, and not a JWT), as specified by the [OAuth 2.0 spec](https://oauth.net/2/access-tokens/). ID tokens are where identity information belongs. That's why we switched from looking for ID information in the access token to looking it up from `/userinfo` when it wasn't available.

**Comment by jsm174** [comment]
@mshima - awesome job! Thank you very much.

@mraible - quick question:

> However, putting identity information in access tokens is also a bad practice.

Makes sense. The main reason I continue to build services with JHipster is the hundreds of eyeballs and best practices. 

Out of the box, JHipster's realm config for keycloak seems to expose everything through the access token:

https://github.com/jhipster/generator-jhipster/blob/a4084738c07aab14ac2fec5e02043a4caef16117/generators/server/templates/src/main/docker/config/realm-config/jhipster-realm.json.ejs#L1328-L1339

Is this something that should probably be looked at because of bad practice?

**Comment by mraible** [comment]
I'm not sure we need to adjust our default Keycloak config because we don't expect people to use it in production. In fact, we don't even generate Keycloak files for Kubernetes. 

I do think it would be useful to create (or link to) a guide for hardening and configuring Keycloak in production.

**Comment by jsm174** [comment]
Cool. 

fwiw, just spent some time removing identity information from Keycloak access tokens. Then I noticed users wouldn't update until server recycled, and then found: https://github.com/jhipster/generator-jhipster/issues/18868.

> NOTE: This is not an issue for regular JHipster apps. It's only an issue when you're using JHipster as an OAuth 2.0 resource server and passing in a JWT access token to authenticate. Currently, this is only done by the Ionic and React Native blueprints. Keycloak will never do the lookup since given_name and family_name are included claims in the access token. For Auth0 and Okta, this auto-lookup makes things easier to configure.

I don't feel so bad now 😄

**Comment by gmarziou** [comment]
Thanks @mraible for the explanation, I did forget about the differences between access and ID tokens, it makes perfect sense now.

**Comment by mshima** [comment]
@DanielFran bounty claimed https://opencollective.com/generator-jhipster/expenses/180801

**Comment by DanielFran** [comment]
@mshima Did not found any bug bounty?

**Comment by mshima** [comment]
@DanielFran not sure I follow.
This bug was fixed by https://github.com/jhipster/generator-jhipster/pull/19454 see https://github.com/jhipster/generator-jhipster/issues/19309#issuecomment-1217376675

**Comment by DanielFran** [comment]
> @DanielFran not sure I follow. This bug was fixed by #19454 see [#19309 (comment)](https://github.com/jhipster/generator-jhipster/issues/19309#issuecomment-1217376675) and [#19309 (comment)](https://github.com/jhipster/generator-jhipster/issues/19309#issuecomment-1217367436)

Sorry, I was in my phone and did not find the labels... 
Already approved
